enzyme/Enzyme/MLIR/Passes/EnzymeBatchDiffPass.cpp

```diff
@@ -36,37 +42,272 @@ namespace mlir {
+
+static Value getBase(Value v) {
+  while (true) {
+    if (auto s = v.getDefiningOp<LLVM::GEPOp>()) {
+      v = s.getBase();
+      continue;
+    }
+    if (auto s = v.getDefiningOp<LLVM::BitcastOp>()) {
+      v = s.getArg();
+      continue;
+    }
+    if (auto s = v.getDefiningOp<LLVM::AddrSpaceCastOp>()) {
+      v = s.getArg();
+      continue;
+    }
+    if (auto s = v.getDefiningOp<memref::CastOp>()) {
+      v = s.getSource();
+      continue;
+    }
+    break;
+  }
+  return v;
+}
+
+static bool isStackAlloca(Value v) {
+  return v.getDefiningOp<memref::AllocaOp>() ||
+         v.getDefiningOp<memref::AllocOp>() ||
+         v.getDefiningOp<LLVM::AllocaOp>();
+}
+
+static bool mayAlias(Value v, Value v2) {
+  v = getBase(v);
+  v2 = getBase(v2);
+  if (v == v2)
+    return true;
+
+  // We may now assume neither v1 nor v2 are subindices
+
+  if (auto glob = v.getDefiningOp<memref::GetGlobalOp>()) {
+    if (auto Aglob = v2.getDefiningOp<memref::GetGlobalOp>()) {
+      return glob.getName() == Aglob.getName();
     }
   }
 
+  if (auto glob = v.getDefiningOp<LLVM::AddressOfOp>()) {
+    if (auto Aglob = v2.getDefiningOp<LLVM::AddressOfOp>()) {
+      return glob.getGlobalName() == Aglob.getGlobalName();
+    }
+  }
+
+  bool isAlloca[2];
+  bool isGlobal[2];
+
+  isAlloca[0] = isStackAlloca(v);
+  isGlobal[0] = v.getDefiningOp<memref::GetGlobalOp>() ||
+                v.getDefiningOp<LLVM::AddressOfOp>();
+
+  isAlloca[1] = isStackAlloca(v2);
+
+  isGlobal[1] = v2.getDefiningOp<memref::GetGlobalOp>() ||
+                v2.getDefiningOp<LLVM::AddressOfOp>();
+
+  // Non-equivalent allocas/global's cannot conflict with each other
+  if ((isAlloca[0] || isGlobal[0]) && (isAlloca[1] || isGlobal[1]))
+    return false;
+
+  bool isArg[2];
+  isArg[0] = isa<BlockArgument>(v) &&
+             isa<FunctionOpInterface>(
+                 cast<BlockArgument>(v).getOwner()->getParentOp());
+
+  isArg[1] = isa<BlockArgument>(v) &&
+             isa<FunctionOpInterface>(
+                 cast<BlockArgument>(v).getOwner()->getParentOp());
+
+  // Stack allocations cannot have been passed as an argument.
+  if ((isAlloca[0] && isArg[1]) || (isAlloca[1] && isArg[0]))
+    return false;
+
+  // Non captured base allocas cannot conflict with another base value.
+  if (isAlloca[0] && !isCaptured(v))
+    return false;
+
+  if (isAlloca[1] && !isCaptured(v2))
+    return false;
+
   return true;
 }
 
+bool mayAlias(MemoryEffects::EffectInstance &a,
+              MemoryEffects::EffectInstance &b,
+              mlir::AliasAnalysis &aliasAnalyzer) {
+  if (a.getResource()->getResourceID() != b.getResource()->getResourceID())
+    return false;
+  Value valA = a.getValue();
+  Value valB = b.getValue();
+
+  // unknown effects may always alias
+  if (!valA || !valB) {
+    return true;
+  }
+  auto valResult = mayAlias(valA, valB);
+  // query alias analysis and polygeist based alias analysis
+  auto aliasResult = aliasAnalyzer.alias(valA, valB);
```

Comment:
@pengmai already has a much more robust alias analysis integrated here (though perhaps doesn't understand the properties above): https://github.com/EnzymeAD/Enzyme/blob/a2af0289d50f4f505f466c10ec488b4f3d7cf77e/enzyme/Enzyme/MLIR/Analysis/DataFlowAliasAnalysis.h